data-analytics/cloud-composer-examples/composer_dataflow_examples/dataflow/process_delimited.py

```diff
@@ -0,0 +1,136 @@
+
+from __future__ import absolute_import
+import argparse
+import logging
+import ntpath
+import re
+
+import apache_beam as beam
+from apache_beam.options import pipeline_options
+
+
+class RowTransformer(object):
+    """A helper class that contains utility methods to parse the delimited file
+    and convert every record into a format acceptable to BigQuery. It also contains
+    utility methods to add a load_dt and a filename fields as demonstration of
+    how records can be enriched as part of the load process."""
+
+    def __init__(self, delimiter, header, filename, load_dt):
+        self.delimiter = delimiter
+
+        # Extract the field name keys from the comma separated input.
+        self.keys = re.split(',', header)
+
+        self.filename_dict = {'filename': filename}
+        self.load_dt_dict = {'load_dt': load_dt}
+
+    def parse(self, row):
+        """This method translates a single delimited record into a dictionary
+        which can be loaded into BigQuery. It also adds filename and load_dt
+        fields to the dictionary."""
+
+        # Strip out the return characters and quote characters.
+        values = re.split(self.delimiter, re.sub('\r\n', '', re.sub(u'"', '', row)))
+
+        row = dict(zip(self.keys, values))
+
+        # Add an additional filename field.
+        row_with_filename = merge_two_dicts(
+            row, self.filename_dict)
+
+        # Add an additional load_dt field.
+        row_with_filename_load_dt = merge_two_dicts(
+            row_with_filename, self.load_dt_dict)
+
+        return row_with_filename_load_dt
+
+
+def merge_two_dicts(first_dict, second_dict):
+    """A utility function to merge two dictionaries."""
+    merged_dict = first_dict.copy()
+    merged_dict.update(second_dict)
+    return merged_dict
+
+
+def run(argv=None):
+    """The main function which creates the pipeline and runs it."""
+    parser = argparse.ArgumentParser()
+
+    # Add the arguments needed for this specific Dataflow job.
+    parser.add_argument(
+        '--input', dest='input', required=True,
+        help='Input file to read.  This can be a local file or '
+             'a file in a Google Storage Bucket.')
+
+    parser.add_argument('--output', dest='output', required=True,
+                        help='Output BQ table to write results to.')
+
+    parser.add_argument('--delimiter', dest='delimiter', required=False,
+                        help='Delimiter to split input records.',
+                        default=',')
+
+    parser.add_argument('--fields', dest='fields', required=True,
+                        help='Comma separated list of field names.')
+
+    parser.add_argument('--load_dt', dest='load_dt', required=True,
+                        help='Load date in YYYY-MM-DD format.')
+
+    known_args, pipeline_args = parser.parse_known_args(argv)
+    parse_delimited = RowTransformer(delimiter=known_args.delimiter,
```

Comment:
You have to provide the header as an option to this script, even though the header is in the input file? I can see that getting out of sync.

Reply:
Agreed! Since this is meant to be an example, I have removed the header from the sample file and only use the header provided as an input.